kedro-org/kedro-viz
# Refactor response classes

## Description

Partially resolves #2061 

- [ ] Module imports related to experiment tracking, strawberry graphQL (2-3sec) - We need to see if we can delay graphql call until someone clicks on ET, remove getVersions call (a middleware which ingests graphQL api app once a request for graphQL is made)
- [ ] Separate api app (main api + additional api) - This will help remove dependencies involved for other apis
- [x] Module imports related to deployer factory (2-3sec) - Deferring deployer factory imports until a /deploy call is made
- [x] Separate responses module and try to get the main route working quickly - This needs refactoring and breaking down our responses module.
- [x] NetworkX takes time for import - Deferring the networkX import until used
- [ ] SQLiteStore initialization also takes time (reason for configure and bootstrap project taking time) - Need to find a way to defer store initialization (likely a North star)
- [x] Apart from all the above improvements, uvicorn.run takes ~10sec to make the app available. Need to experiment by making a lighter app, reducing api routes and dependencies.

## Development notes

* Split `responses.py` file and group the response classes under `responses` folder which helps in scaling and maintaining the responses code base
* Defer heavy imports like `NetworkX`, deployer factory imports until used
* By experimenting found the real reason for `uvicorn.run` taking time is app building which involves importing all kedro project modules (example - for demo project, sklearn, matplotlib etc take time as shown below).
* Updated tests and import references

<img width="2046" alt="image" src="https://github.com/user-attachments/assets/5512eb61-9c5f-45d1-b300-1e65f334ab16">
 
**Next steps:**

* Moving away from multiprocessing to threading
* https://github.com/kedro-org/kedro-viz/issues/2079

## QA notes

* All tests should pass
* Since the responses module is refactored, there needs to be manual testing for confirming the CLI and UI responses. It would be nice if we test `kedro viz build`, `kedro viz deploy` are intact
* @jitu5 I think Kedro VSCode uses a function `get_kedro_project_json_data` in responses file which is moved to `pipelines.py` file. I would request you to have a look at this

## Checklist
